database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -161,23 +185,33 @@ private boolean isKeyAtIndex(String key, int index) {
         return index >= 0 && index < size() && mDataSnapshots.get(index).getKey().equals(key);
     }
 
-    protected void onKeyAdded(DataSnapshot data) {
+    /**
+     * @deprecated
+     */
+    @Deprecated
+    protected void onKeyAdded(DataSnapshot data) {}
+
+    protected void onKeyAdded(DataSnapshot data, int index) {
+        onKeyAdded(data);
```

Comment:
Unfortunately, this isn't backwards compatible if people call super on the old method. 😢 I totally understand if you don't want to proceed with this change though we would be missing out on some significant perf gains: O(n) with the old method, O(1) with the new method on simple additions.